components/tidb_query_datatype/src/codec/collation/utf8mb4.rs

```diff
@@ -395,6 +396,111 @@ impl Collator for CollatorUtf8Mb4GeneralCi {
     }
 }
 
+#[inline]
+fn unicode_ci_convert(c: char) -> u128 {
+    let r = c as usize;
+    if r > 0xFFFF {
+        return 0xFFFD;
+    }
+
+    let u = UNICODE_CI_TABLE[r];
+    if u == LONG_RUNE {
+        return map_long_rune(r);
+    }
+
+    u as u128
+}
+
+/// Collator for `utf8mb4_unicode_ci` collation with padding behavior (trims right spaces).
+#[derive(Debug)]
+pub struct CollatorUtf8Mb4UnicodeCi;
+
+impl Collator for CollatorUtf8Mb4UnicodeCi {
+    type Charset = CharsetUtf8mb4;
+
+    #[inline]
+    fn validate(bstr: &[u8]) -> Result<()> {
+        str::from_utf8(bstr)?;
+        Ok(())
+    }
+
+    fn write_sort_key<W: BufferWriter>(writer: &mut W, bstr: &[u8]) -> Result<usize> {
+        let s = str::from_utf8(bstr)?.trim_end_matches(TRIM_PADDING_SPACE);
+        let mut n = 0;
+        for ch in s.chars() {
+            let mut convert = unicode_ci_convert(ch);
+            while convert != 0 {
+                writer.write_u16_be((convert & 0xFFFF) as u16)?;
+                n += 1;
+                convert >>= 16
+            }
+        }
+        Ok(n * std::mem::size_of::<u16>())
+    }
+
+    fn sort_compare(a: &[u8], b: &[u8]) -> Result<Ordering> {
+        let mut ca = str::from_utf8(a)?
+            .trim_end_matches(TRIM_PADDING_SPACE)
+            .chars();
+        let mut cb = str::from_utf8(b)?
+            .trim_end_matches(TRIM_PADDING_SPACE)
+            .chars();
+        let mut an = 0;
+        let mut bn = 0;
+
+        loop {
+            if an == 0 {
+                for ach in &mut ca {
```

Comment:
I'm not 100% sure I understand this. What is the for loop for? Is it jumping overall `\0`? 'cause I don't see a similar loop in https://github.com/pingcap/tidb/blob/03d4ae0d9dac199bfe512f1cc4f61b5811dbde6e/util/collate/unicode_ci.go#L97-L104. Are they equivalent?

Reply:
they are equivalent. character's weight is `0` in `unicode_ci` is ignorable, should be ignore. https://github.com/pingcap/tidb/blob/03d4ae0d9dac199bfe512f1cc4f61b5811dbde6e/util/collate/unicode_ci.go#L42-L54 do this job. `next` function will only `return` if not `0` or string is done.